lib/rucio/daemons/storage/consistency/actions.py

```diff
@@ -0,0 +1,805 @@
+    labels = {'rse': rse}
+
+    record_gauge('storage.consistency.actions_dark_files_found',
+      confirmed_dark_files, labels=labels)
+    record_gauge('storage.consistency.actions_dark_files_confirmed',
+      confirmed_dark_files, labels=labels)
+
+    deleted_files = 0
+    if confirmed_dark_files/max_files_at_site < maxdarkfraction\
+      or force_proceed is True:
+        logger(logging.INFO, '\n Can proceed with dark files deletion')
+
+# Then, do the real deletion (code from DeleteReplicas.py)
+        issuer = InternalAccount('root')
+        with open(confirmed_dark, 'r') as csvfile:
+            reader = csv.reader(csvfile)
+            dark_replicas = []
+            #for rse, scope, name, reason in reader:
+            scope = "cms"
+            reason = "deleteing dark file"
+            for name, in reader:
+                logger(logging.INFO, '\n Processing a dark file:\n RSE %s  Scope: %s  Name: %s'\
+                 % (rse, scope, name))
+                rse_id = get_rse_id(rse=rse)
+                Intscope = InternalScope(scope=scope, vo=issuer.vo)
+                lfns = [{'scope': scope, 'name': name}]
+
+                attributes = get_rse_info(rse=rse)
+                pfns = lfns2pfns(rse_settings=attributes, lfns=lfns, operation='delete')
+                pfn_key = scope + ':' + name
+                url = pfns[pfn_key]
+                urls = [url]
+                paths = parse_pfns(attributes, urls, operation='delete')
+                replicas = [{'scope': Intscope, 'rse_id': rse_id, 'name': name,
+                  'path': paths[url]['path']+paths[url]['name']}]
+                add_quarantined_replicas(rse_id, replicas, session=None)
+                deleted_files += 1
+                labels = {'rse': rse}
+                record_counter('storage.consistency.actions_dark_files_deleted_counter',
+                  delta=1, labels=labels)
+
+        #Update the stats
+        t1 = time.time()
+
+        cc_stats.update({
+            "end_time": t1,
+            "initial_dark_files": dark_files,
+            "confirmed_dark_files": deleted_files,
+            "status": "done"
+        })
+        stats[stats_key] = cc_stats
+        record_gauge('storage.consistency.actions_dark_files_deleted',
+          deleted_files, labels=labels)
+
+    else:
+        darkperc = 100.*confirmed_dark_files/max_files_at_site
+        logger(logging.WARNING, '\n ATTENTION: Too many DARK files! (%3.2f%%) \n\
+         Stopping and asking for operators help.' % darkperc)
+
+        #Update the stats
+        t1 = time.time()
+
+        cc_stats.update({
+            "end_time": t1,
+            "initial_dark_files": dark_files,
+            "confirmed_dark_files": 0,
+            "status": "ABORTED",
+            "aborted_reason": "%3.2f%% dark" % darkperc,
+        })
+        stats[stats_key] = cc_stats
+        record_gauge('storage.consistency.actions_dark_files_deleted', 0, labels=labels)
```

Comment:
A lot of this lines are duplicated. Consider refactoring it

Reply:
This can be addressed in the final version of the Stats class, when Igor contributes it separately to Rucio.
Placing a TODO: tag here to remind us of that fact.